core/host.go

```diff
@@ -5,122 +5,54 @@ import (
 	"fmt"
 	"path/filepath"
 	"strings"
-	"sync"
 
-	bkclient "github.com/moby/buildkit/client"
 	"github.com/moby/buildkit/client/llb"
-	bkgw "github.com/moby/buildkit/frontend/gateway/client"
-	"github.com/moby/buildkit/solver/pb"
 	specs "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
+type Host struct {
+	Workdir string
+}
+
+func NewHost(workdir string) *Host {
+	return &Host{
+		Workdir: workdir,
+	}
+}
+
 type HostVariable struct {
 	Name string `json:"name"`
 }
 
-type HostDirectory struct {
-	ID HostDirectoryID `json:"id"`
-}
+func (host *Host) Directory(ctx context.Context, dirPath string, platform specs.Platform) (*Directory, error) {
+	var absPath string
+	var err error
+	if filepath.IsAbs(dirPath) {
+		absPath = dirPath
```

Comment:
I think even if it's absolute, we should probably run an EvalSymlinks on dirPath, or at least .Clean. Just so sychronization of the same directory specified different ways can be deduped. Doesn't handle every case (i.e. bind mounts and such) but is simple enough that we may as well include it as a nicety I think.